File: claude-code/seeds/doc-extract.js
// Seed: doc-extract — extract a fixed schema from documents, no silent drops.
// Claude Code Workflow edition. args: { docs: [{id, text}], fields: ["title","date",...] }
export const meta = {
  name: 'doc-extract',
  description: 'Extract structured fields from documents against a fixed schema; report anything that failed.',
  phases: [{ title: 'Extract' }],
}

const AGENT_TIMEOUT_MS = 240000  // guard: a hung agent resolves null, never stalls parallel()
const aw = (p, o) => Promise.race([
  agent(p, o),
  new Promise((r) => setTimeout(() => { log(`timeout: ${(o && o.label) || 'agent'}`); r(null) }, AGENT_TIMEOUT_MS)),
])

const docs = (args && args.docs) || []
const fields = (args && args.fields) || ['title', 'date', 'summary']
const schema = {
  type: 'object', additionalProperties: false, required: fields,
  properties: Object.fromEntries(fields.map((f) => [f, { type: 'string' }])),
}

phase('Extract')
const got = await parallel(docs.map((d) => () =>
  aw('Extract exactly these fields. Use "" if a field is genuinely absent — never guess or ' +
        `fabricate:\n${JSON.stringify(fields)}\n\nDOCUMENT:\n${d.text}`,
        { label: `extract:${d.id}`, phase: 'Extract', model: 'sonnet', schema })
    .then((r) => ({ id: d.id, fields: r }))))

const records = got.filter((g) => g && g.fields)
const missing = docs.filter((d, i) => !(got[i] && got[i].fields)).map((d) => d.id)
if (missing.length) log(`WARNING: no extraction for ${JSON.stringify(missing)} (reported, not silently dropped)`)
return { records, missing, extracted: records.length, total: docs.length }

File: seeds/doc-extract.py
"""Seed: doc-extract — extract a fixed schema from documents, no silent drops.

args: {"docs": [{"id","text"}, ...], "fields": ["title","date","summary", ...]}
"""

META = {
    "name": "doc-extract",
    "description": "Extract structured fields from documents against a fixed schema.",
    "phases": ["Extract"],
}


async def run(agent, parallel, pipeline, log, phase, args):
    a = args or {}
    docs = a.get("docs", [])                       # [{"id","text"}]
    fields = a.get("fields", ["title", "date", "summary"])
    schema = {
        "type": "object", "additionalProperties": False,
        "properties": {f: {"type": "string"} for f in fields},
        "required": fields,
    }

    async def extract(d):
        r = await agent(
            "Extract exactly these fields. Use \"\" if a field is genuinely absent — never "
            f"guess or fabricate:\n{fields}\n\nDOCUMENT:\n{d['text']}",
            label=f"extract:{d['id']}", phase="Extract", model="sonnet", schema=schema)
        return {"id": d["id"], "fields": r}

    got = await parallel([(lambda d=d: extract(d)) for d in docs])
    records = [g for g in got if g and g.get("fields")]
    missing = [d["id"] for d, g in zip(docs, got) if not (g and g.get("fields"))]
    if missing:
        log(f"WARNING: no extraction for {missing} (reported, not silently dropped)")
    return {"records": records, "missing": missing,
            "extracted": len(records), "total": len(docs)}
